client/starwhale/api/_impl/dataset/builder.py

```diff
@@ -48,10 +43,212 @@
 _header_size = _header_struct.size
 _header_version = 0
 
-
 _BDType = t.TypeVar("_BDType", bound="BaseBuildExecutor")
 
 
+class BinWriter(Protocol):
+    total_bin_size: int
+
+    def write_row(self, row: TabularDatasetRow) -> None:
+        """
+        Find large bytes or local fs file in row data. Convert them to accessible link
+        """
+        ...
+
+    def flush(self) -> None:
+        ...
+
+    def __enter__(self) -> BinWriter:
+        return self
+
+    def __exit__(
+        self,
+        type: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        trace: TracebackType,
+    ) -> None:
+        ...
+
+
+class FragmentBinWriter:
+
+    """
+    bin format:
+        header_magic    uint32  I
+        crc             uint32  I
+        _reserved       uint64  Q
+        size            uint32  I
+        padding_size    uint32  I
+        header_version  uint32  I
+        data_magic      uint32  I --> above 32 bytes
+        data bytes...
+        padding bytes...        --> default 4K padding
+    """
+
+    class _BinSection(t.NamedTuple):
+        offset: int
+        size: int
+        raw_data_offset: int
+        raw_data_size: int
+
+    class _SrcPathSpec(t.NamedTuple):
+        src_path: Path
+        remove_src: bool
+
+    def __init__(
+        self,
+        work_dir: Path,
+        data_output_dir: Path,
+        tabular_dataset: TabularDataset,
+        alignment_bytes_size: int = D_ALIGNMENT_SIZE,
+        volume_bytes_size: int = D_FILE_VOLUME_SIZE,
+        bin2link_threshold: int = D_BIN_TO_LINK_THRESHOLD,
+    ) -> None:
+        self.tabular_dataset = tabular_dataset
+        self.work_dir = work_dir
+        self.data_output_dir = data_output_dir
+        self.volume_bytes_size = volume_bytes_size
+        self.bin2link_threshold = bin2link_threshold
+        self.alignment_bytes_size = alignment_bytes_size
+        self.ds_copy_candidates: t.Dict[str, FragmentBinWriter._SrcPathSpec] = {}
+        self.wrote_size = 0
+        _, bin_writer_path = tempfile.mkstemp(
+            prefix="bin-writer-", dir=str(self.work_dir.absolute())
+        )
+        self.dwriter_path = Path(bin_writer_path)
+        self.dwriter = self.dwriter_path.open("wb")
+        self.total_bin_size = 0
+        self._to_update_rows: t.List[TabularDatasetRow] = []
+        self._lock = threading.Lock()
+
+    def __enter__(self) -> FragmentBinWriter:
+        return self
+
+    def __exit__(
+        self,
+        type: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        trace: TracebackType,
+    ) -> None:
+        if value:  # pragma: no cover
+            logger.warning(f"type:{type}, exception:{value}, traceback:{trace}")
+
+        self.close()
+
+    def write_row(self, row: TabularDatasetRow) -> None:
```

Comment:
can we split `FragmentBinWriter` into two classes by `artifacts_with_bin` feature?

Reply:
> can we split `FragmentBinWriter` into two classes by `artifacts_with_bin` feature?

let's split it in the future?

Reply:
ok